File: controller/e_rainbow_shift.cpp
//
// Created by Nico Wittig on 2018-12-19.
//

#include "e_rainbow_shift.h"

e_rainbow_shift::e_rainbow_shift(light_element *le) : effect(le, EFFECT_NAME) {
}

e_rainbow_shift::~e_rainbow_shift() {
}

bool e_rainbow_shift::init() {

    for (int led = 0; led < le->get_num_leds(); led++) {
        le->leds[led] = CHSV(led * 255/le->get_num_leds(), 255, 255);
    }

    return true;
}

bool e_rainbow_shift::run() {
    shift(1);
    return false;
}

File: server/e_rainbow_shift.py
from color import Color
from effect import Effect


class E_Rainbow_Shift(Effect):
    def __init__(self, light_element, soft_shift=False, shift_forward=True, invert_colors=False, speed=1):
        super().__init__(light_element, "Rainbow Shift")

        self.__soft_shift = soft_shift
        self.__shift_forward = shift_forward
        self.__invert_colors = invert_colors
        self.__speed = abs(speed)

    def init(self):
        if self.__invert_colors:
            super().set_color(Color((255,255,255)), Color((0,255,255)), hue_cycle_forward=False)
        else:
            super().set_color(Color((0,255,255)), Color((255,255,255)), hue_cycle_forward=True)
        return True

    def run(self):
        if self.__soft_shift:
            super().increase_hue(self.__speed if self.__shift_forward else -self.__speed)
        else:
            super().shift(self.__speed if self.__shift_forward else -self.__speed)
        return True
